flink-formats/flink-compress/src/main/java/org/apache/flink/formats/compress/CompressWriterFactory.java

```diff
@@ -42,33 +47,41 @@
 
 	private Extractor<IN> extractor;
 	private CompressionCodec hadoopCodec;
+	private String hadoopCodecName;
+	private Map<String, String> hadoopConfigurationMap;
 
 	public CompressWriterFactory(Extractor<IN> extractor) {
 		this.extractor = Preconditions.checkNotNull(extractor, "extractor cannot be null");
+		this.hadoopConfigurationMap = new HashMap<>();
 	}
 
 	public CompressWriterFactory<IN> withHadoopCompression(String hadoopCodecName) {
 		return withHadoopCompression(hadoopCodecName, new Configuration());
 	}
 
 	public CompressWriterFactory<IN> withHadoopCompression(String hadoopCodecName, Configuration hadoopConfiguration) {
-		return withHadoopCompression(new CompressionCodecFactory(hadoopConfiguration).getCodecByName(hadoopCodecName));
-	}
+		this.hadoopCodecName = hadoopCodecName;
+
+		for (Map.Entry<String, String> entry : hadoopConfiguration) {
+			hadoopConfigurationMap.put(entry.getKey(), entry.getValue());
+		}
 
-	public CompressWriterFactory<IN> withHadoopCompression(CompressionCodec hadoopCodec) {
```

Comment:
why you remove this method?

in this moment is the only way to use compression with some hack
( extending the codec and implementing serializable )

Reply:
I believe that's not a clean approach.

Reply:
I aggre with you, its not a clean solution, and its not a solution for this problem.

but this method allow you to pass directly the codec (if you have it) or a implementation of it (if need to implement it)

Reply:
Even if it is a custom implementation, providing the complete class name, simple name or even alias should just be fine. You just have to add your custom implementation's name in Hadoop configuration for the property `io.compression.codecs` in either Hadoop's Configuration file or by programatically.

Hadoop's `CompressionCodecFactory` takes care of the rest using `ServiceLoader`. I have added test cases to cover that.

Reply:
perfect.